shell/platform/darwin/macos/framework/Source/FlutterMenuPlugin.mm

```diff
@@ -0,0 +1,420 @@
+      // If we're adding the first menu in the menu bar, then just set the items
+      // in the submenu, rather than adding a new menu. This is because AppKit
+      // gets very squirrelly when you remove the <APP_NAME> menu, so we just
+      // empty it out in removeAllMenus, and refill it here. This ignores the
+      // label of the first item (because it will always be set to the APP_NAME
+      // by AppKit).
+      NSMenu* appNameMenu = [[NSApp.mainMenu itemAtIndex:0] submenu];
+      for (int i = 0; i < [menuItem.submenu numberOfItems]; ++i) {
+        [appNameMenu addItem:[[menuItem.submenu itemAtIndex:i] copy]];
+      }
+      firstMenu = false;
+    } else {
+      [NSApp.mainMenu addItem:menuItem];
+    }
+    NSNumber* boxedID = item[kIdKey];
+    FlutterMenuDelegate* delegate = [[FlutterMenuDelegate alloc] initWithId:boxedID.longLongValue
+                                                                 andChannel:_channel];
+    [_menuDelegates addObject:delegate];
+    [[NSApp.mainMenu itemAtIndex:index++] submenu].delegate = delegate;
+  }
+}
+
+- (NSMenuItem*)findDefaultMenu:(NSMenu*)menu ofType:(SEL)selector withTag:(int)tag {
+  auto enumerator = menu != nullptr ? menu.itemArray.objectEnumerator
+                                    : _platformProvidedItemArray.objectEnumerator;
+  for (NSMenuItem* item in enumerator) {
+    if ([item action] == selector && (tag == 0 || [item tag] == tag)) {
+      return item;
+    }
+    if ([[item submenu] numberOfItems] > 0) {
+      NSMenuItem* child_find = [self findDefaultMenu:[item submenu] ofType:selector withTag:tag];
+      if (child_find != nullptr) {
+        return child_find;
+      }
+    }
+  }
+  return nullptr;
+}
+
+- (NSMenuItem*)createPlatformProvidedMenu:(PlatformProvidedMenu)type {
+  NSMenuItem* found = nullptr;
+  auto found_type = kMacOSProvidedMenus.find(type);
+  if (found_type == kMacOSProvidedMenus.end()) {
+    return nullptr;
+  }
+  SEL selector_target = found_type->second;
+  // Since it doesn't have a definitive selector, the Services submenu is
+  // assumed to be the first item with a submenu action in the first menu item
+  // of the default menu set. We can't just get the title to check, since that
+  // is localized.
+  NSMenu* starting_menu = type == PlatformProvidedMenu::servicesSubmenu
+                              ? [_platformProvidedItemArray[0] submenu]
+                              : nullptr;
+  found = [self findDefaultMenu:starting_menu ofType:selector_target withTag:0];
+  if (found != nullptr) {
+    return [found copy];
+  }
+  return nullptr;
+}
+
+/**
+ * Constructs and returns an NSMenuItem corresponding to the item in
+ * |representation|, including recursively creating children if it has a
+ * submenu.
+ */
+- (NSMenuItem*)menuItemFromFlutterRepresentation:(NSDictionary*)representation {
+  if ([(NSNumber*)([representation valueForKey:kDividerKey]) intValue] == YES) {
+    return [NSMenuItem separatorItem];
+  } else {
```

Comment:
Optional nit: Using an "else" after an early return needlessly indents everything.

(Personally, I think it also makes the logic harder to follow since it makes it look like a hybrid of early return and return-once-at-the end, even though it's actually just the former.)

Reply:
Yeah, this was just an oversight.  I agree with you.